core/src/unprocessed_packet_batches.rs

```diff
@@ -26,14 +25,62 @@ pub struct DeserializedPacket {
     is_simple_vote: bool,
 }
 
+/// Defines the type of entry in `UnprocessedPacketBatches`, it holds original packet_batch
+/// for forwarding, as well as `forwarded` flag;
+/// Each packet in packet_batch are deserialized upon receiving, the result are stored in
+/// `DeserializedPacket` in the same order as packets in `packet_batch`.
 #[derive(Debug, Default)]
 pub struct DeserializedPacketBatch {
     pub packet_batch: PacketBatch,
     pub forwarded: bool,
-    // indexes of valid packets in batch, and their corrersponding deserialized_packet
+    // indexes of valid packets in batch, and their corresponding deserialized_packet
     pub unprocessed_packets: HashMap<usize, DeserializedPacket>,
 }
 
+/// Currently each banking_stage thread has a `UnprocessedPacketBatches` buffer to store
+/// Packet_batches received from sigverify. Banking thread continuously scans the buffer
+/// to pick proper packets to add to block.
+#[derive(Default)]
+pub struct UnprocessedPacketBatches(VecDeque<DeserializedPacketBatch>);
+
+impl std::ops::Deref for UnprocessedPacketBatches {
+    type Target = VecDeque<DeserializedPacketBatch>;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl std::ops::DerefMut for UnprocessedPacketBatches {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl RetainMut<DeserializedPacketBatch> for UnprocessedPacketBatches {
+    fn retain_mut<F>(&mut self, f: F)
+    where
+        F: FnMut(&mut DeserializedPacketBatch) -> bool,
+    {
+        RetainMut::retain_mut(&mut self.0, f);
+    }
+}
```

Comment:
Two questions around this, since I haven't played with `RetainMut`.
1. Since `UnprocessedPacketBatches` implements `DerefMut`, would it be sufficient to impl `RetainMut` for `DeserializedPacketBatch` instead?
2. If (1) is a 'no, can the actual impl instead be:
```rust
RetainMut::retain_mut(&mut self, f);
```
(notice without the `.0`)
Again, since `UnprocessedPacketBatches` implements `DerefMut`.